src/pbkdf2.rs

```diff
@@ -1,9 +1,175 @@
 // Copyright 2024 Simo Sorce
 // See LICENSE.txt file for terms
 
+use super::attribute;
+use super::err_rv;
+use super::error;
+use super::hmac;
+use super::interface;
+use super::mechanism;
+use super::object;
+
+use attribute::{from_bool, from_bytes, from_ulong};
+use error::{KError, KResult};
+use interface::*;
+use mechanism::*;
+use object::{Object, ObjectFactories};
+
+use super::bytes_to_vec;
+
+use std::fmt::Debug;
+
+#[cfg(not(feature = "fips"))]
 use std::mem::swap;
 
+pub fn register(mechs: &mut Mechanisms, _: &mut ObjectFactories) {
+    PBKDF2Mechanism::register_mechanisms(mechs);
+}
+
+#[derive(Debug)]
+struct PBKDF2Mechanism {
+    info: CK_MECHANISM_INFO,
+}
+
+impl PBKDF2Mechanism {
+    fn register_mechanisms(mechs: &mut Mechanisms) {
+        mechs.add_mechanism(
+            CKM_PKCS5_PBKD2,
+            Box::new(PBKDF2Mechanism {
+                info: CK_MECHANISM_INFO {
+                    ulMinKeySize: 0,
+                    ulMaxKeySize: std::u32::MAX as CK_ULONG,
+                    flags: CKF_GENERATE,
+                },
+            }),
+        );
+    }
+
+    fn mock_password_object(&self, key: Vec<u8>) -> KResult<Object> {
+        let mut obj = Object::new();
+        obj.set_zeroize();
+        obj.set_attr(from_ulong(CKA_CLASS, CKO_SECRET_KEY))?;
+        obj.set_attr(from_ulong(CKA_KEY_TYPE, CKK_GENERIC_SECRET))?;
+        obj.set_attr(from_ulong(CKA_VALUE_LEN, key.len() as CK_ULONG))?;
+        obj.set_attr(from_bytes(CKA_VALUE, key))?;
+        obj.set_attr(from_bool(CKA_DERIVE, true))?;
+        Ok(obj)
+    }
+}
+
+impl Mechanism for PBKDF2Mechanism {
+    fn info(&self) -> &CK_MECHANISM_INFO {
+        &self.info
+    }
+
+    fn generate_key(
+        &self,
+        mech: &CK_MECHANISM,
+        template: &[CK_ATTRIBUTE],
+        mechanisms: &Mechanisms,
+        objfactories: &ObjectFactories,
+    ) -> KResult<Object> {
+        if self.info.flags & CKF_GENERATE != CKF_GENERATE {
+            return err_rv!(CKR_MECHANISM_INVALID);
+        }
+        if mech.mechanism != CKM_PKCS5_PBKD2 {
+            return err_rv!(CKR_MECHANISM_INVALID);
+        }
+
+        if mech.ulParameterLen as usize
+            != ::std::mem::size_of::<CK_PKCS5_PBKD2_PARAMS2>()
+        {
+            return err_rv!(CKR_ARGUMENTS_BAD);
+        }
+        let params =
+            unsafe { *(mech.pParameter as *const CK_PKCS5_PBKD2_PARAMS2) };
```

Comment:
use `cast_params` here .

Reply:
cast_params was introduced in a later patch and I did not want to cause too much churn in the code.

Are you suggesting I introduce it and change all places for consistency ?

Reply:
Yes please. Not urgent though. I wanted to use it in a new code (eddsa) and then realized that its not yet merged. Searching for the other uses, I realized how ugly the original code without macros is and how much nicer it would be without that.

Reply:
ok I will do it now, or I'll forget, expect a few more patches to land in this PR